README.txt
.. contents:: Table of Contents
   :depth: 2

Products.ZPyODBCDA
****************************************

Overview
--------

ZPyODBCDA is a Zope Database Adapter based on Python's pyodbc moudle 
(http://pyodbc.sourceforge.net). It links the Zope server to ODBC data sources. 
It is distributed under the GPL licence.

The reason I wrote a new Zope ODBC Database Adapter and didn't simply use the
existing "New ZODBC DA" on zope.org is that I felt the underlying Pywin32's 
ODBC module which the "New ZODBC DA" depended on was not stable enough, 
and occasionally crashed my Python programs.

The mxODBCDA is a commercial product which is not free.

The Products.ZPyODBCDA product is modified from the codes of Thierry MICHEL's 
GPL licensed ZPoPyDA product, which links Zope to PostgreSQL and seems to be 
very mature.

Products.ZPyODBCDA requires Python 2.4 or higher.

If your Zope runs on the win32 platform, you probably don't need any other moudules, as I already included pyodbc's binary code (pyodbc.pyd) in the package folder; for other platform like Linux, you may first need to install the pyodbc moudule for your Python instance.

I hope this product can provide a more reliable free Zope/ODBC connect solution for you.


Requirements
------------

    * Pyodbc (http://code.google.com/p/pyodbc/)
    
    * Zope 2.12
    
    * Plone 4 (optional)
    
Installation
------------
    
To enable this product,on a buildout based installation:

    1. Edit your buildout.cfg and add ``Products.ZPyODBCDA``
       to the list of eggs to install ::

        [buildout]
        ...
        eggs = 
            Products.ZPyODBCDA
    
After updating the configuration you need to run the ''bin/buildout'',
which will take care of updating your system.

Go to the 'Site Setup' page in the Plone interface and click on the
'Add/Remove Products' link.

Choose the product (check its checkbox) and click the 'Install' button.

Uninstall -- This can be done from the same management screen, but only
if you installed it from the quick installer.

Note: You may have to empty your browser cache and save your resource registries
in order to see the effects of the product installation.

Sponsoring
----------

Adaptation of this product was sponsored by:

    * `TRT13 <http://www.trt13.jus.br/>`_
    * `Simples Consultoria <http://www.simplesconsultoria.com.br/>`_.


Credits
-------
    
    * Henry Zhou (Jiangwen365 at gmail dot com) - Coding
    * Simples Consultoria (products at simplesconsultoria dot com dot br) - 
      Egg implementation
    







CHANGES.rst
1.0.9.3+perfact.7
=================
- Drop support for Python 2
- Remove version pinning on pyodbc

1.0.9.3+perfact.6
=================
- Close connection in rollback exception so it's reopenend later

1.0.9.3+perfact.5
=================

- Fix versioning to be PEP 440 compliant

1.0.9.3.perfact5
================

- Raise an error if the result set of any query exceeds the maximum number of
  rows configured in either the connector or the query properties.

1.0.9.3.perfact4
================

- Drop separate escaping routine so the inherited one from ``Connection`` is
  used. In particular, escaping of apostrophes was broken and did not follow
  the SQL standard of doubling the apostrophe, instead trying to escape with a
  backslash.

1.0.9.3.perfact3
================

- Fix escaping routine. Since the transition to allow usage in both Python 2
  and 3, the joining tried to use a list as joiner, which does not work.


Products/ZPyODBCDA/__init__.py
#
# Copyright (C) 2007 Henry Zhou <jiangwen365@gmail.com>
# Copyright (C) 2001 Thierry MICHEL <thierry@nekhem.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
#

"""Z PyODBC Database Adapter Package Registration"""

import sys
import traceback

from .pyodbc_browser import pyodbcConnectionBrowser
from .pyodbc_browser import addpyodbcConnectionBrowser
from .pyodbc_DA import addpyodbcConnectionForm


def initialize(context):
    "Initialize the product."
    try:
        context.registerClass(
            pyodbcConnectionBrowser,
            constructors=(
                addpyodbcConnectionForm,
                addpyodbcConnectionBrowser
            ),
            icon='www/pyodbc_icon.png'
        )
    except Exception:
        type, val, tb = sys.exc_info()
        sys.stderr.write(
            traceback.format_exception(
                type, val, tb
            ).join('')
        )
        del type, val, tb


Products/ZPyODBCDA/pyodbc_db.py
#
# Copyright (C) 2008 Henry Zhou <jiangwen365@gmail.com>
# Copyright (C) 2001 Thierry MICHEL <thierry@nekhem.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
#


import pyodbc
import locale
import os
import sys
import datetime
from DateTime import DateTime
from Shared.DC.ZRDB.TM import TM

if sys.version_info[1] <= 4:
    pyodbc_dir = os.path.dirname(os.path.realpath(__file__)) + '\\pyodbc24'
else:
    pyodbc_dir = os.path.dirname(os.path.realpath(__file__)) + '\\pyodbc26'
sys.path.append(pyodbc_dir)

# Patch JH 08/2016: Decimals are formatted according to the locale set
# at "import" time.
orig_locale = locale.getlocale(locale.LC_NUMERIC)
locale.setlocale(locale.LC_NUMERIC, 'C')
# End of Patch JH 08/2016


# Patch JH 08/2016: Restore original locale.
locale.setlocale(locale.LC_NUMERIC, orig_locale)
# End of Patch JH 08/2016

sys.path.remove(pyodbc_dir)


DB_Error = pyodbc.Error


def convert9(value):
    # Data Type 9, not recognized by ODBC
    # is a Null
    return None


class DB(TM):
    """DB class from pyodbc driver"""

    def __init__(self, po_conx):
        conn_param = eval(po_conx)
        self._conxString = conn_param['connx_string']
        self._auto_commit = conn_param['auto_commit']
        self._MaxRows = conn_param['MaxRows']

        self._conx = pyodbc.connect(self._conxString,
                                    autocommit=self._auto_commit)
        # Here we add a new data type converter
        self._conx.add_output_converter(9, convert9)

        self._cursor = self._conx.cursor
        self._numTry = 0
        self._numMaxTry = 5

    def setAutocommit(self, auto_commit=None):
        self._auto_commit = auto_commit

    def setMaxRows(self, MaxRows):
        try:
            self._MaxRows = int(MaxRows)
        except ValueError as mesg:
            raise mesg

    def connect(self, ps_queryString):
        from time import sleep
        mesg = ''
        while (self._numTry < self._numMaxTry):
            self._numTry = self._numTry + 1
            try:
                self._conx = pyodbc.connect(self._conxString,
                                            self._auto_commit)
                self._cursor = self._conx.cursor
                return self.query(ps_queryString)
            except pyodbc.OperationalError as err:
                mesg = err
                sleep(5)
        self._numTry = 0
        raise mesg

    def close(self):
        pass

    def query(self, ps_queryString, pl_maxRows=None):
        self._register()
        # This seems not necessary and sometimes breaks certain SQLs.
        # ps_queryString = ps_queryString.replace("\n"," ")
        # pyodbc version 4 needs to receive unicode, not bytes.
        if (pyodbc.version.split('.')[0] == '4' and
                isinstance(ps_queryString, bytes)):
            ps_queryString = ps_queryString.decode('utf-8')
        try:
            o_cur = self._cursor()
            o_cur.execute(ps_queryString)
            # Occasionally the connection is lost which I haven't figured
            # out why, but this re-connecting works for me very well.
            # Need to improve though!
        except pyodbc.Error:
            try:
                o_cur.close()
                self._conx.close()
            except Exception:
                pass
            self._conx = pyodbc.connect(self._conxString,
                                        autocommit=self._auto_commit)
            self._cursor = self._conx.cursor
            o_cur = self._cursor()
            o_cur.execute(ps_queryString)

        try:
            while True:
                o_desc = o_cur.description

                if o_desc:
                    max_rows = (pl_maxRows if pl_maxRows is not None
                                else self._MaxRows)
                    if max_rows:
                        o_result = o_cur.fetchmany(max_rows)
                        # Do not truncate a query result if it exceeds the
                        # maximum row number set in either the connector or
                        # query properties.
                        if len(o_result) == max_rows:
                            try:
                                overshoot_result = o_cur.fetchone()
                            except Exception:
                                overshoot_result = None
                            if overshoot_result:
                                assert False, (
                                    "This query has returned more than "
                                    "MaxRows results. Please raise "
                                    "MaxRows or limit in SQL.")
                    else:
                        o_result = o_cur.fetchall()
                if not o_cur.nextset():
                    break
        finally:
            o_cur.close()

        if o_desc is None:
            return (), ()

        o_items = [
            {
                'name': x[0],
                'type': x[1],
                'dsize': x[2],
                'isize': x[3],
                'precision': x[4],
                'scale': x[5],
                'null': x[6]
            } for x in o_desc
        ]

        field_types = [i['type'] for i in o_items]

        # get date field ids to be convert to DateTime type
        date_field_ids = [i for i in range(len(field_types))
                          if field_types[i] in (datetime.datetime,
                                                datetime.date)
                          ]
        # get float field ids to be convert from decimal to float
        float_field_ids = [i for i in range(len(field_types))
                           if field_types[i] in (float,)
                           ]

        # if any pyodbc's object type needs to be converted to
        # Zope's object type.
        if len(date_field_ids) + len(float_field_ids) > 0:
            for row in o_result:
                for field_id in date_field_ids:
                    if row[field_id] is not None:
                        # Currently we don't do timezones. Everything is UTC.
                        # Ideally we'd get the current Oracle timezone
                        # and use that.
                        row[field_id] = DateTime(
                            *(row[field_id].timetuple()[:6] + ('UTC',)))
                for field_id in float_field_ids:
                    if row[field_id] is not None:
                        row[field_id] = float(row[field_id])

        return o_items, o_result

    def _datetime_convert(self, dt, val):
        if dt and (val is not None):
            # Currently we don't do timezones. Everything is UTC.
            # Ideally we'd get the current Oracle timezone and use that.
            x = val.timetuple()[:6] + ('UTC',)
            return DateTime(*x)
        return val

    def _Decimal_convert(self, dt, val):
        if dt and (val is not None):
            return float(val)
        return val

    def getdbinfo(self):
        dbinfo = ''
        try:
            dbinfo += self._conx.getinfo(pyodbc.SQL_DBMS_NAME) + ' '
            dbinfo += self._conx.getinfo(pyodbc.SQL_DBMS_VER) + ' '
            dbinfo += self._conx.getinfo(pyodbc.SQL_DATABASE_NAME) + ' '
        except Exception:
            pass
        return dbinfo

    def get_columns(self, table_name):
        o_cur = self._cursor()
        columns = [column for column in o_cur.columns(table_name)]
        o_cur.close()
        return columns

    def get_tables(self):
        o_cur = self._cursor()
        tables = [table for table in o_cur.tables()]
        o_cur.close()
        return tables

    def get_primaryKeys(self, table_name):
        o_cur = self._cursor()
        primaryKeys = []
        try:
            primaryKeys = [p_key for p_key in o_cur.primaryKeys(table_name)]
        except Exception:
            pass
        o_cur.close()
        return primaryKeys

    def get_foreignKeys(self, table_name):
        o_cur = self._cursor()
        foreignKeys = []
        try:
            foreignKeys = [f_key for f_key in o_cur.foreignKeys(
                foreignTable=table_name)]
        except Exception:
            pass
        o_cur.close()
        return foreignKeys

    def _finish(self):
        try:
            self._conx.commit()
        except DB_Error:
            raise

    def _abort(self):
        try:
            self._conx.rollback()
        except DB_Error:
            self._conx.close()

    def _begin(self):
        pass


Products/ZPyODBCDA/pyodbc_browser.py
#
# Copyright (C) 2007 Henry Zhou <jiangwen365@gmail.com>
# Copyright (C) 2001 Thierry MICHEL <thierry@nekhem.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
#
# Here are requests to have informations about tables.

from .pyodbc_DA import pyodbcConnection
from .pyodbc_DA import DTMLFile


def addpyodbcConnectionBrowser(self, id, title, connection_string,
                               auto_commit=None, connected=None, MaxRows=10,
                               REQUEST=None):
    """Add a Z PyODBC DB connection to a folder"""

    if auto_commit is not None:
        auto_commit = 1
    if connected is not None:
        connected = 1
    self._setObject(id, pyodbcConnectionBrowser(id,
                                                title,
                                                connection_string,
                                                auto_commit,
                                                connected,
                                                MaxRows))
    if REQUEST is not None:
        return self.manage_main(self, REQUEST)


class pyodbcConnectionBrowser(pyodbcConnection):
    """pyodbc Connection with tables browser feature"""

    manage_options = pyodbcConnection.manage_options + (
        {'label': 'Browser', 'action': 'manage_browser'},
    )

    manage_browser = DTMLFile('dtml/pyodbcconnectionBrowser', globals())

    def __is_foreign_key(self, description):
        res = 1
        if description.find('FOREIGN') == -1:
            res = 0
        return res

    def __get_source_table(self, arguments):
        res = ''
        res = arguments.split('\\000')[2]

        return res

    def __get_source_field(self, arguments):
        res = arguments.split('\\000')[4]
        return res

    def __set_table_fkeys(self, tablename):
        if self._v_connected:
            records = self._v_database_connection.get_foreignKeys(tablename)
        self.fkeys = []
        for r in records:
            # (source table, source column)
            self.fkeys.append(((r[6], r[7]), r[2]))

    def __set_table_pkeys(self, tablename):
        if self._v_connected:
            records = self._v_database_connection.get_primaryKeys(tablename)
        self.pkeys = []
        for r in records:
            self.pkeys.append((r[2], r[3]))

    def is_primary_key(self, tablename, field):
        if (tablename, field) in self.pkeys:
            return 1
        return 0

    def is_foreign_key(self, tablename, field):
        for fkey in self.fkeys:
            if (tablename, field) == fkey[0]:
                return fkey[1]
        return 0

    def manage_get_table_infos(self, tablename):
        " "
        if self._v_connected:
            column_infos = self._v_database_connection.get_columns(tablename)
            self.__set_table_pkeys(tablename)
            self.__set_table_fkeys(tablename)
        else:
            return []

        result = []
        i = 1
        for info in column_infos:
            rsp = {}
            rsp['attnum'] = i
            rsp['attname'] = info[3]
            rsp['typname'] = info[5]
            rsp['attnotnull'] = info[10]
            rsp['attlen'] = info[6]
            result.append(rsp)
            i = i + 1
        return result

    def manage_get_tables(self):
        " "
        if self._v_connected:
            r = self._v_database_connection.get_tables()
        else:
            return []

        result = []
        for item in r:
            rsp = {}
            rsp['tablename'] = item[2]
            rsp['tabletype'] = item[3]
            result.append(rsp)

        return result


Products/ZPyODBCDA/pyodbc_DA.py
#
# Copyright (C) 2007 Henry Zhou <jiangwen365@gmail.com>
# Copyright (C) 2001 Thierry MICHEL <thierry@nekhem.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
#


from Shared.DC.ZRDB.Connection import DTMLFile, Connection
from .pyodbc_db import DB

database_type = 'PyODBC'

"""%s Database Connection""" % database_type


addpyodbcConnectionForm = DTMLFile(
    'dtml/pyodbcconnectionAdd',
    globals(),
    default_id='%s_database_connection' % database_type,
    default_title='Z %s Database Connection' % database_type,
    database_type=database_type,
)


def addpyodbcConnection(self, id, title, connection_string, auto_commit=None,
                        connected=None, MaxRows=10, REQUEST=None):
    """Add a Z PyODBC DB connection to a folder"""

    if connected is not None:
        connected = 1
    self._setObject(id, pyodbcConnection(id,
                                         title,
                                         connection_string,
                                         auto_commit,
                                         connected,
                                         MaxRows))
    if REQUEST is not None:
        return self.manage_main(self, REQUEST)


class pyodbcConnection(Connection):
    """A database connection object"""
    database_type = database_type
    id = '%s_database_connection' % database_type
    meta_type = title = 'Z %s Database Connection' % database_type
    icon = 'misc_/Z%sDA/conn' % database_type
    _isAnSQLConnection = 1

    def __init__(self, id, title, connection_string, auto_commit=None,
                 connected=None, MaxRows=10):
        self.connx_string = connection_string
        self.MaxRows = MaxRows
        if auto_commit is not None:
            self.auto_commit = True
        else:
            self.auto_commit = False

        conn_param = {}
        conn_param['connx_string'] = self.connx_string
        conn_param['auto_commit'] = self.auto_commit
        conn_param['MaxRows'] = self.MaxRows

        Connection.__init__(self, id, title, repr(conn_param), connected)

    def get_backend_info(self):
        if self._v_connected:
            return self._v_database_connection.getdbinfo()
        return ''

    def manage_MaxRows(self, MaxRows, REQUEST):
        " "
        self.MaxRows = MaxRows
        self._v_database_connection.setMaxRows(MaxRows)
        return self.manage_main(self, REQUEST)

    manage_main = DTMLFile('dtml/pyodbcconnectionStatus', globals())

    manage_properties = DTMLFile('dtml/pyodbcconnectionEdit', globals())

    def manage_edit(self, title, connection_string, auto_commit=None,
                    connected=None, MaxRows=10, REQUEST=None):
        """Change connection"""
        self.connx_string = connection_string
        self.MaxRows = MaxRows
        if auto_commit is not None:
            self.auto_commit = True
        else:
            self.auto_commit = False

        conn_param = {}
        conn_param['connx_string'] = self.connx_string
        conn_param['auto_commit'] = self.auto_commit
        conn_param['MaxRows'] = self.MaxRows
        Connection.manage_edit(self, title, repr(
            conn_param), connected, REQUEST)

    def factory(self):
        return DB
